### PR TITLE
LOVE-912 Merge Basic EKS Blueprints

### PR DESCRIPTION
**Pull request by erincinci**
## Definition of Done

This is the definition of done for (new and modified) blueprints to be accepted into the [central XebiaLabs blueprints repository](https://github.com/xebialabs/blueprints).

### Value

This section describes how to determine whether a blueprint has enough value to be included in the central XebiaLabs blueprints repository.

- [x] The blueprint applies to a focused use case.
- [x] The blueprint uses at least one XebiaLabs product.
- [x] The blueprint depends on as few other products as possible. **N.B.:** This is part of the definition of done to keep the blueprint focused and applicable to as many users as possible. A blueprint that depends on one specific CI tool **and** one specific Java EE application server **and** one specific ticketing system can only be used by users that have all three products.
- [x] The blueprint can be used with content supplied by the user but also includes demo content. A blueprint that does not work with content supplied by the user is a demo or a showcase and does not help the user with their own work.

### Technical

This section describes how to determine whether the blueprint has the right technical quality to be included in the central XebiaLabs blueprints repository.

- [x] The branch from which the PR is created is up-to-date with the `development` branch.
- [x] The blueprint contains a `README.md` file at the root of the blueprint folder that describes the blueprint, using the README template in the `.github` folder of this repository.
- [x] The blueprint generates a `xebialabs/USAGE.md` file which, at a minimum, explains how to use the blueprint after it is generated (like adding any missing steps, creating accounts, setting up Docker containers, applying the YAML using `xl` CLI, running release. etc.). **N.B.:** Do not use this document to describe how to instantiate the blueprint. It will only be available to the user *after* the blueprint has been instantiated.
- [x] If the blueprint depends on external products that can be started as a Docker container, the blueprint generates a `docker/docker-compose.yml` file so that it can be tested without having to manually install those products. Do not use this Docker Compose file to start XL Deploy, XL Release, Docker or Kubernetes.
- [x] The blueprint does not contain sensitive information such passwords, tokens, credentials or licenses.
- [x] The blueprint uses `secret: true` in the `blueprint.yaml` parameter definition for question that ask for sensitive information.
- [x] The blueprint does not contain the files `xebialabs/.gitignore`, `xebialabs/values.xlvals` and `xebialabs/secrets.xlvals` and does not refer to them from the files section of the `blueprint.yaml` file. These files will be generated when the blueprint is instantiated. To generate the `xebialabs/values.xlvals` file, use the `saveInXlVals: true` directive in the parameters section of the blueprint. To generate the `xebialabs/secrets.xlvals` file, use the `secret: true` directive .
- [x] The blueprint does not define parameters for trivial things like phase names, task names, folder names etc. Ask questions that add value and be opinionated where possible. For example, ask for a project name and derive folder names, task names etc from that.
- [x] Folder and file names must use [kebab-case](http://wiki.c2.com/?KebabCase), for example: `aws/sample-app-demo`, `xld-environment.yaml`

### Review and testing

This section describes how to determine whether the blueprint has been reviewed and tested well enough to be included in the central XebiaLabs blueprints repository.

- [x] Unit test is added in a `__test__` folder for each blueprint. Refer the `CONTRIBUTING.md` file at the root of of this repository for more details.
- [x] The Travis CI for the PR is green
- [x] The blueprint has been reviewed by someone else in the team.
- [x] Both README and USAGE files have been reviewed by a technical writer and a product marketing manager.
- [x] The blueprint has been manually tested with the `docker/docker-compose.yml` that is generated as part of it.
- [ ] The blueprint works on Linux, Mac and Windows.

### CODE REVIEW

File: aws/basic-eks-cluster/blueprint.yaml [hunk 1]
```diff
@@ -123,6 +124,29 @@ spec:
...
+  - name: AddIAMUser
+    description: "Do you want to add extra user to have an access to EKS via kubectl"
+    type: Confirm
+    default: true
+
+  - name: EKSUserARN
+    description: "Enter existent IAM user ARN to be used to access EKS cluster via kubectl"
+    type: Input
+    default: "arn:aws:iam::100000000000:user/username"
+    pattern: ^arn:aws:iam::\d{12}:(?:root|user\/[A-Za-z0-9]+)$
```
Comment (ikarlashov) [inline aws/basic-eks-cluster/blueprint.yaml]: This pattern doesn't support dashes in "username" part. 
Value arn:aws:iam::932770550094:user/test-eks was not validated by the patter though it's a valid AWS IAM ARN.

File: aws/basic-eks-cluster/kubernetes/aws-auth-cm.yaml [hunk 2]
```diff
@@ -10,8 +10,10 @@ data:
       groups:
         - system:bootstrappers
         - system:nodes
+  {{if .AddIAMUser }}
   mapUsers: |
-    - userarn: ##EksUserArn##
-      username: ##EksUser##
+    - userarn: ##EKS_USER_ARN##
```
Comment (ikarlashov) [inline aws/basic-eks-cluster/kubernetes/aws-auth-cm.yaml]: These variable names are incorrect. Actually previous CamelCase version was right. In dictionary u have `EksUser: '{{"{{%outputVariables.EksUser%}}"}}'
EksUserArn: '{{"{{%outputVariables.EksUserArn%}}"}}`

**Comment by erincinci** [comment]
> I got the error:
> 
> `[ERROR]: 2019-05-15 14:34:18.960000+00:00: CREATE_FAILED for CFNSecretProvider. Reason: 1 validation error detected: Value &apos;test-eks-blueprint-cfn-secret-provider-test-eks-blueprint-eks-user&apos; at &apos;functionName&apos; failed to satisfy constraint: Member must have length less than or equal to 64 (Service: AWSLambdaInternal; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: 859c0725-771e-11e9-a34e-1d164ddee872) [ERROR]: 2019-05-15 14:34:19.893000+00:00: CREATE_FAILED for test-eks-blueprint-eks-user. Reason: The following resource(s) failed to create: [CFNSecretProvider]. `
> 
> Seems we need to limit lambda function name?

Limiting appName to 16 chars, according to the AWS docs function name can be 64 chars max, and because we have suffix/prefix to it it would be calculated to have 16 max as safe limit.
https://docs.aws.amazon.com/lambda/latest/dg/API_CreateFunction.html#API_CreateFunction_RequestSyntax

File: aws/basic-eks-cluster/blueprint.yaml (hunk 1 above)
Comment (erincinci) [reply aws/basic-eks-cluster/blueprint.yaml]: Updated regex pattern

File: aws/basic-eks-cluster/kubernetes/aws-auth-cm.yaml (hunk 2 above)
Comment (erincinci) [reply aws/basic-eks-cluster/kubernetes/aws-auth-cm.yaml]: fixed

File: aws/basic-eks-cluster/blueprint.yaml [hunk 3]
```diff
@@ -66,6 +67,7 @@ spec:
   - name: NodeInstanceType
     type: Select
     description: What EC2 instance type do you want to use for the nodes?
+    default: t2.xlarge
```
Comment (deepu105) [inline aws/basic-eks-cluster/blueprint.yaml]: This is quite big to be the default. IMO we should have `t2.medium` as default.

File: aws/basic-eks-cluster/blueprint.yaml [hunk 4]
```diff
@@ -123,6 +125,29 @@ spec:
     - p3.8xlarge
     - p3.16xlarge
 
+  - name: AddIAMUser
+    description: "Do you want to add extra user to have an access to EKS via kubectl"
+    type: Confirm
+    default: true
+
+  - name: EKSUserARN
+    description: "Enter existent IAM user ARN to be used to access EKS cluster via kubectl"
```
Comment (deepu105) [inline aws/basic-eks-cluster/blueprint.yaml]:
```suggestion
    description: "Enter existing IAM user ARN to be used to access EKS cluster via kubectl"
```

File: aws/basic-eks-cluster/blueprint.yaml [hunk 5]
```diff
@@ -123,6 +125,29 @@ spec:
...
+    type: Input
+    default: "arn:aws:iam::100000000000:user/username"
+    pattern: ^arn:aws:(?:iam|sts)::\d{12}:(?:[A-Za-z0-9-_\/\(\)]+)$
+    dependsOnTrue: AddIAMUser
+
+  - name: EKSUserName
+    description: "Enter existent IAM username to access EKS cluster via kubectl"
```
Comment (deepu105) [inline aws/basic-eks-cluster/blueprint.yaml]:
```suggestion
    description: "Enter existing IAM username to access EKS cluster via kubectl"
```

File: aws/basic-eks-cluster/blueprint.yaml [hunk 6]
```diff
@@ -123,6 +125,29 @@ spec:
...
 
+  - name: AddIAMUser
+    description: "Do you want to add extra user to have an access to EKS via kubectl"
+    type: Confirm
+    default: true
+
+  - name: EKSUserARN
+    description: "Enter existent IAM user ARN to be used to access EKS cluster via kubectl"
+    type: Input
+    default: "arn:aws:iam::100000000000:user/username"
```
Comment (deepu105) [inline aws/basic-eks-cluster/blueprint.yaml]: Is this a valid ARN?

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl [hunk 7]
```diff
@@ -110,15 +110,11 @@ Parameters:
   NodeAutoScalingGroupMinSize:
     Type: Number
     Description: Minimum size of Node Group ASG.
-    Default: 1
+    Default: "{{.NodeAutoScalingGroupDesiredSize}}"
```
Comment (deepu105) [inline aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: This was being passed down via the XLD template already so why should this be added here?

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl [hunk 8]
```diff
@@ -49,7 +49,7 @@ Parameters:
   NodeInstanceType:
     Description: EC2 instance type for the node instances
     Type: String
-    Default: t2.medium
+    Default: "{{.NodeInstanceType}}"
```
Comment (deepu105) [inline aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: This is already passed via XLD

File: aws/basic-eks-cluster/xebialabs/xld-cloudformation-apps.yaml.tmpl [hunk 9]
```diff
@@ -101,9 +101,6 @@ spec:
             ClusterStackName: {{$app}}-eks-master
             ClusterName: {{$app}}-master
             NodeGroupName: {{$app}}
-            NodeInstanceType: {{.NodeInstanceType}}
```
Comment (deepu105) [inline aws/basic-eks-cluster/xebialabs/xld-cloudformation-apps.yaml.tmpl]: why this change?

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl [hunk 10]
```diff
@@ -110,15 +110,11 @@ Parameters:
...
+    Default: "{{.NodeAutoScalingGroupDesiredSize}}"
   NodeAutoScalingGroupMaxSize:
     Type: Number
     Description: Maximum size of Node Group ASG.
-    Default: 4
-  NodeAutoScalingGroupDesiredSize:
-    Type: Number
-    Description: Desired size of Node Group ASG.
-    Default: 3
+    Default: "{{.NodeAutoScalingGroupDesiredSize}}"
```
Comment (deepu105) [inline aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: why are these changed, EKS doesn't let you have the same min/max in all region and NodeInstanceType, for some combinations EKS will error saying min/max can not be same, that's why I was adding +1 to the desiredSize for max

Comment (deepu105) [reply aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: also we are not utilizing the autoscale capacity if we set the same value for all

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl [hunk 11]
```diff
@@ -16,6 +15,36 @@ spec:
     - {{$app}}
     scriptUsername: !value XL_RELEASE_USERNAME
     scriptUserPassword: !value XL_RELEASE_PASSWORD
+
+    variables:     
+      - key: AWSaccessKey
```
Comment (deepu105) [inline aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: This seems unnecessary

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl [hunk 12]
```diff
@@ -16,6 +15,36 @@ spec:
...
+        type: xlrelease.StringVariable
+        requiresValue: true
+        showOnReleaseStart: false
+        value: {{.AWSAccessKey}}
+      - key: AWSaccessSecret
```
Comment (deepu105) [inline aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: same here seems unnecessary

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl [hunk 13]
```diff
@@ -16,6 +15,36 @@ spec:
...
+        type: xlrelease.StringVariable
+        requiresValue: true
+        showOnReleaseStart: false
+        value: {{.AWSAccessSecret}}
+      - key: AWSRegion
```
Comment (deepu105) [inline aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: unnecessary

File: aws/basic-eks-cluster/cloudformation/efs.yaml [hunk 14]
```diff
@@ -1,13 +1,12 @@
...
+Description: >
+  This template deploys an EFS file system.
 
 Parameters:
   EnvironmentName:
     AllowedPattern: ^[0-9a-zA-Z]+([0-9a-zA-Z-]*[0-9a-zA-Z])*$
     Description: An environment name that will be prefixed to resource names.
     Type: String
-    Default: "xl-cluster"
+    Default: "xl-up"
```
Comment (deepu105) [inline aws/basic-eks-cluster/cloudformation/efs.yaml]: why xl-up?

File: aws/basic-eks-cluster/blueprint.yaml [hunk 15]
```diff
@@ -123,6 +125,29 @@ spec:
     - p3.8xlarge
     - p3.16xlarge
 
+  - name: AddIAMUser
```
Comment (deepu105) [inline aws/basic-eks-cluster/blueprint.yaml]: why are we asking these questions? we are provisioning this anyway right?

Comment (deepu105) [reply aws/basic-eks-cluster/blueprint.yaml]: also these 3 variables are not used anywhere?

Comment (eltonkevani) [reply aws/basic-eks-cluster/blueprint.yaml]: nope. only if someone want to add a user he can, we do not provision  users anymore

File: aws/basic-eks-cluster/cloudformation/efs.yaml (hunk 14 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/cloudformation/efs.yaml]: because is cool

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl (hunk 13 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: not if the user did not use .aws/credentials

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl (hunk 12 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: not if the user did not use .aws/credentials

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl (hunk 11 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: not if the user did not use .aws/credentials

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl (hunk 10 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: let the user to decide how many nodes he wants

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl (hunk 7 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: test it

File: aws/basic-eks-cluster/blueprint.yaml (hunk 6 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/blueprint.yaml]: nope just example. can be put on description but then will be too long

File: aws/basic-eks-cluster/blueprint.yaml (hunk 3 above)
Comment (eltonkevani) [reply aws/basic-eks-cluster/blueprint.yaml]: we help aws to make money

Comment (deepu105) [reply aws/basic-eks-cluster/blueprint.yaml]: stop giving them money

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl (hunk 7 above)
Comment (deepu105) [reply aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: I tested it already, wasn't it working?

File: aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl (hunk 10 above)
Comment (deepu105) [reply aws/basic-eks-cluster/cloudformation/eks-workers.yaml.tmpl]: that is the desired capacity we let people choose, also on EKS it makes sense to use autoscaling IMO, also the fact having same min/max doesnt work for some region/NodeInstanceType

File: aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl (hunk 11 above)
Comment (deepu105) [reply aws/basic-eks-cluster/xebialabs/xlr-pipeline-ci-cd.yaml.tmpl]: These are not used anywhere so what is the purpose of this here?

File: aws/basic-eks-cluster/cloudformation/efs.yaml (hunk 14 above)
Comment (deepu105) [reply aws/basic-eks-cluster/cloudformation/efs.yaml]: malaka

File: aws/basic-eks-cluster/blueprint.yaml (hunk 15 above)
Comment (deepu105) [reply aws/basic-eks-cluster/blueprint.yaml]: look at the code dude, this is not used anywhere

Comment (eltonkevani) [reply aws/basic-eks-cluster/blueprint.yaml]: https://github.com/xebialabs/blueprints/blob/qpi-1/aws/eks-provisioning/kubernetes/aws-auth-cm.yaml.tmpl something probably went wrong with the merge

Comment (deepu105) [reply aws/basic-eks-cluster/blueprint.yaml]: i'm gonna clean this shit up, we dont want too many questions here anyway @vpartington wants these basic ones to be simple

Comment (eltonkevani) [reply aws/basic-eks-cluster/blueprint.yaml]: up to you
